File: src/543_Diameter_of_Binary_Tree.py
# 543_Diameter_of_Binary_Tree
# https://leetcode.cn/problems/diameter-of-binary-tree/

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:        
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def depth(node: TreeNode) -> int:
            if not node:
                return 0 
            left_depth = depth(node.left)
            right_depth = depth(node.right)
            self.ans = max(self.ans, left_depth + right_depth)
            return max(left_depth, right_depth) + 1
        
        self.ans = 0
        depth(root)
        return self.ans
        

File: src/543_Diameter_of_Binary_Tree.rs
// 543. 二叉树的直径
// https://leetcode.cn/problems/diameter-of-binary-tree/

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}

use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn diameter_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut ans = 0;
        depth(&root, &mut ans);
        ans
    }
}

pub fn depth(node: &Option<Rc<RefCell<TreeNode>>>, ans: &mut i32) -> i32 {
    if let Some(n) = node {
        let left_depth = depth(&n.borrow().left, ans);
        let right_depth = depth(&n.borrow().right, ans);
        *ans = (*ans).max(left_depth + right_depth);
        left_depth.max(right_depth) + 1
    } else {
        0
    }
}